add linux.psscan

eve-mem opened the pull request:
This adds psscan for linux, based on vol2 plugin by @atcuno - https://github.com/volatilityfoundation/volatility/blob/master/volatility/plugins/linux/psscan.py

Example output:
```
python vol.py -f linux-sample-1.dmp linux.psscan
Volatility 3 Framework 2.4.1
Progress:  100.00               Stacking attempts finished
OFFSET (P)      PID     TID     PPID    COMM

0x1ef5140       6949    6949    8503
0x4404100       3929    3929    1       goa-daemon
0x4404800       8577    8577    0       sed
0x44a5180       3929    3934    1       gdbus
0x44a5880       8600    8600    8599    bash
0x1ac980c0      2390    2390    2388    avahi-daemon
0x1ac987c0      2363    2363    1       NetworkManager
0x1ac9c040      2363    2386    1       gdbus
0x1ac9c740      8503    8503    8502    bash
0x1ac9e1c0      2393    2396    1       gdbus
0x1ac9e8c0      3885    3885    3754    gnome-fallback-
0x1aca0080      2344    2344    1       atd
0x1aca0780      2532    2532    1       cron
0x1aca5180      2393    2393    1       polkitd
0x1aca5880      2652    2660    1       console-kit-dae
0x1acc2100      2425    2425    1       gdm3
0x1acc2800      2425    2443    1       gmain
0x1ad0e100      2803    2803    1       smbd
<snip>
```

eve-mem commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+            kernel_add = 0
+        else:
+            pack_format = "Q"
+            kernel_add = 0xffff000000000000
```
This wont work as I intended all the time for the reasons listed in this issue (PML4 v LA57) - It should also probably be an 'or' rather than an 'add' -  https://github.com/volatilityfoundation/volatility3/issues/883

Any advice welcome, its to ensure that the bytes scanned for in the physical layer match the actual pointers we're looking for. Rather than missing the first `0xffff` off.

paulkermann replied:
#702 at your service.

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+# This file is Copyright 2021 Volatility Foundation and licensed under the Volatility Software License 1.0
```
Sorry, it's extremely minor, but please could you update the copyright year to the year it gets contributed during which you contribute it?

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+    def _get_task_fields(
+        self, task: interfaces.objects.ObjectInterface
+    ) -> Tuple[int, int, int, str]:
+        """Extract the fields needed for the final output
```
Awesome!!!  5:D  We love docstrings, thanks!  5:D

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+        Returns:
+            A tuple with the fields to show in the plugin output.
+        """
+        pid = task.tgid
```
These seemed to be named... confusingly.  I don't know the linux structures well enough, so I'm going to assume this is right, but just wanted you to check to make sure it wasn't a typo please.  5:)

eve-mem replied:
Yes it is weird, its to match the pslist plugin (https://github.com/volatilityfoundation/volatility3/blob/develop/volatility3/framework/plugins/linux/pslist.py#L83) . 

It's the way the kernel replies to the getpid() function itself - https://github.com/torvalds/linux/blob/master/kernel/sys.c#L932. 

This stackoverflow reply has a nice explanation too - https://stackoverflow.com/questions/9305992/if-threads-share-the-same-pid-how-can-they-be-identified

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+        """Generates the tasks found from scanning."""
+
+        for task in self.scan_tasks(
+            self.context, self.config["kernel"], self.config["kernel.symbol_table_name"]
```
This is... kinda okay, but if the key isn't present or the kernel module is generated in some other way, then `kernel.symbol_table_name` might not be what we're after.  I think the correct method for getting this value is to get the kernel from the list, and then ask it from which symbol table name it has.  Sorry, I know it's a very corner case, but I'm not sure we can strictly rely on that key always existing, only the direct requirements on this plugin or attributes of objects that fulfill the direct requirements of this plugin.

eve-mem replied:
I think I've now done this in a better way - what do you think?

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,127 @@
+                                    layer_name='memory_layer')
+            
+            # sanity check exit_state
+            if ptask.exit_state not in [0, 16, 32, 16|32]:
```
It would be good to document what these flags are for and/or turn them into symbols or at least an enum, so the names can be referenced and people understand which exit_states are being filtered).

eve-mem replied:
Yes, that's a great idea. I've included it in the output too now.

ikelos reviewed (commented):
Thanks just a few comments from me (and the usual sign-off needed by @atcuno ), but looking good!  5:)

eve-mem commented:
I think I've updated this now to take into account all of your comments @ikelos. I really liked the idea of including the exit state more verbosely, as you're likely running a psscan plugin to find dead processes.

I also added debug logging to the sanity checks as it felt useful. 

Here is an updated example output:

```
python vol.py -f linux-sample-1.dmp  linux.psscan
Volatility 3 Framework 2.4.1
Progress:  100.00               Stacking attempts finished
OFFSET (P)      PID     TID     PPID    COMM    EXIT_STATE

0x1ef5140       6949    6949    8503            EXIT_ZOMBIE
0x4404100       3929    3929    1       goa-daemon      TASK_RUNNING
0x4404800       8577    8577    0       sed     EXIT_ZOMBIE
0x44a5180       3929    3934    1       gdbus   TASK_RUNNING
0x44a5880       8600    8600    8599    bash    TASK_RUNNING
0x1ac980c0      2390    2390    2388    avahi-daemon    TASK_RUNNING
0x1ac987c0      2363    2363    1       NetworkManager  TASK_RUNNING
0x1ac9c040      2363    2386    1       gdbus   TASK_RUNNING
0x1ac9c740      8503    8503    8502    bash    TASK_RUNNING
0x1ac9e1c0      2393    2396    1       gdbus   TASK_RUNNING
0x1ac9e8c0      3885    3885    3754    gnome-fallback- TASK_RUNNING
0x1aca0080      2344    2344    1       atd     TASK_RUNNING
0x1aca0780      2532    2532    1       cron    TASK_RUNNING
0x1aca5180      2393    2393    1       polkitd TASK_RUNNING
0x1aca5880      2652    2660    1       console-kit-dae TASK_RUNNING
0x1acc2100      2425    2425    1       gdm3    TASK_RUNNING
0x1acc2800      2425    2443    1       gmain   TASK_RUNNING
0x1ad0e100      2803    2803    1       smbd    TASK_RUNNING
0x1ad0e800      3824    3824    1       pulseaudio      TASK_RUNNING
0x1ad160c0      3830    3832    1       gdbus   TASK_RUNNING
0x1ad167c0      2652    2661    1       console-kit-dae TASK_RUNNING
0x1ad46180      3807    3815    3754    dconf worker    TASK_RUNNING
<snip>
```

eve-mem commented:
I've updated this now to use @ikelos [canonical-helper](https://github.com/volatilityfoundation/volatility3/pull/882/commits/c1e425217bf8ea0e4b62a56ddaff5db29d87b4f0) feature which is nice.

atcuno commented:
@ikelos this looks good to me

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,158 @@
+        """Generates the tasks found from scanning."""
+
+        for task in self.scan_tasks(
+            self.context, self.config["kernel"], self.config["kernel.layer_name"]
```
This still uses `kernel.layer_name`, sorry to be a pain, but to ensure things don't break in future it would be better to pull out the kernel and then ask it for the layer_name associated with it.

ikelos commented on volatility3/framework/plugins/linux/psscan.py:
```diff
@@ -0,0 +1,158 @@
+                continue
+
+            # sanity check pid
+            if not (0 < ptask.pid < 66000):
```
Is this checking that pid should be less than 65535?  I know it's just a safety check, but if there's a specific limit, I'd prefer to use that than something that's "close"...

ikelos reviewed (commented):
This looks really good, could I ask you to rebase it please, so the change itself applies cleanly to `develop`, and that merge won't contain changes to the intel layer directly?  I believe they're already part of `develop`, so the rebase should do this all for you.  After that it should be good to merge!  5:D

eve-mem commented:
Hello @ikelos I'm it sure that the intel layer updates have been merged into develop yet. Unless im missing something? Sorry to ask but would you mind checking and correctly me if i am wrong.

ikelos commented:
You're quite right, they hadn't gone in yet, they were just sat in a branch, but I'd like to merge them as their own PR, rather than including them in this.  It makes it easier to keep change separate and in small chunks.

I've now made the PR as #905, and I'll update the various locations mentioning this issue, and then merge it after the weekend, so there's time for people to review/comment on it.  After that's done, this can be rebased against it and merged in.  I hope that's all ok, sorry for the long route to getting it in, but I hope it'll prevent issues down the road...  5:)

eve-mem commented:
That sounds great. I'll make those other changes and keep an eye out.

eve-mem committed:
Update linux.psscan

eve-mem committed:
Update linux.psscan to find kernel layer name correctly

ikelos reviewed (approved):
Looks good, thanks!  5:D

eve-mem committed:
Update linux.psscan with black linting and version
